manager/orchestrator/global/global_test.go

```diff
@@ -267,3 +267,197 @@ func deleteTask(t *testing.T, s *store.MemoryStore, task *api.Task) {
+			NodeID:    "nodeid1",
+		},
+	}
+	for _, task := range tasks {
+		addTask(t, s, task)
+	}
+
+	// watch orchestration events
+	watch, cancel := state.Watch(s.WatchQueue(), state.EventCreateTask{}, state.EventUpdateTask{}, state.EventDeleteTask{})
+	defer cancel()
+
+	orchestrator := NewGlobalOrchestrator(s)
+	defer orchestrator.Stop()
+
+	go func() {
+		assert.NoError(t, orchestrator.Run(ctx))
+	}()
+
+	observedTask1 := testutils.WatchTaskUpdate(t, watch)
+	assert.Equal(t, observedTask1.ID, "task1")
+	assert.Equal(t, observedTask1.Status.State, api.TaskStateFailed)
+	assert.Equal(t, observedTask1.DesiredState, api.TaskStateRunning)
+
+	observedTask2 := testutils.WatchTaskUpdate(t, watch)
+	assert.Equal(t, observedTask2.ID, "task1")
+	assert.Equal(t, observedTask2.Status.State, api.TaskStateFailed)
+	assert.Equal(t, observedTask2.DesiredState, api.TaskStateShutdown)
+
+	observedTask3 := testutils.WatchTaskCreate(t, watch)
+	assert.Equal(t, observedTask3.NodeID, "nodeid1")
+	assert.Equal(t, observedTask3.Status.State, api.TaskStateNew)
+	assert.Equal(t, observedTask3.DesiredState, api.TaskStateReady)
+}
+
+func TestInitializationExtraTask(t *testing.T) {
+	ctx := context.Background()
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+	defer s.Close()
+
+	// create nodes, services and tasks in store directly
+	// where orchestrator runs, it should fix tasks to declarative state
+	addNode(t, s, node1)
+	addService(t, s, service1)
+	tasks := []*api.Task{
+		// nodeid1 has 2 tasks for serviceid1
+		{
+			ID:           "task1",
+			Slot:         0,
+			DesiredState: api.TaskStateRunning,
+			Status: api.TaskStatus{
+				State: api.TaskStateRunning,
+			},
+			Spec: api.TaskSpec{
+				Runtime: &api.TaskSpec_Container{
+					Container: &api.ContainerSpec{},
+				},
+			},
+			ServiceAnnotations: api.Annotations{
+				Name: "task1",
+			},
+			ServiceID: "serviceid1",
+			NodeID:    "nodeid1",
+		},
+		{
+			ID:           "task2",
+			Slot:         0,
+			DesiredState: api.TaskStateRunning,
+			Status: api.TaskStatus{
+				State: api.TaskStateRunning,
+			},
+			Spec: api.TaskSpec{
+				Runtime: &api.TaskSpec_Container{
+					Container: &api.ContainerSpec{},
+				},
+			},
+			ServiceAnnotations: api.Annotations{
+				Name: "task2",
+			},
+			ServiceID: "serviceid1",
+			NodeID:    "nodeid1",
+		},
+	}
+	for _, task := range tasks {
+		addTask(t, s, task)
+	}
+
+	// watch orchestration events
+	watch, cancel := state.Watch(s.WatchQueue(), state.EventCreateTask{}, state.EventUpdateTask{}, state.EventDeleteTask{})
+	defer cancel()
+
+	orchestrator := NewGlobalOrchestrator(s)
+	defer orchestrator.Stop()
+
+	go func() {
+		assert.NoError(t, orchestrator.Run(ctx))
+	}()
+
+	observedTask1 := testutils.WatchTaskUpdate(t, watch)
+	assert.True(t, observedTask1.ID == "task1" || observedTask1.ID == "task2")
+	assert.Equal(t, observedTask1.Status.State, api.TaskStateRunning)
+	assert.Equal(t, observedTask1.DesiredState, api.TaskStateShutdown)
```

Comment:
Maybe we should do a `View` after this and make sure we see only one task with a desired state between "ready" and "running".